mlflow/pyfunc/model.py

```diff
@@ -0,0 +1,275 @@
+import os
+import tempfile
+import shutil
+import sys
+import yaml
+from abc import ABCMeta, abstractmethod
+
+import cloudpickle
+
+import mlflow.pyfunc
+import mlflow.utils
+from mlflow.exceptions import MlflowException
+from mlflow.models import Model
+from mlflow.protos.databricks_pb2 import INVALID_PARAMETER_VALUE, RESOURCE_ALREADY_EXISTS
+from mlflow.tracking.utils import _download_artifact_from_uri
+from mlflow.utils.environment import _mlflow_conda_env
+from mlflow.utils.model_utils import _get_flavor_configuration
+from mlflow.utils.file_utils import TempDir, _copy_file_or_tree
+
+DEFAULT_CONDA_ENV = _mlflow_conda_env(
+    additional_conda_deps=[
+        "cloudpickle=={}".format(cloudpickle.__version__),
+    ],
+    additional_pip_deps=None,
+    additional_conda_channels=None,
+)
+
+CONFIG_KEY_ARTIFACTS = "artifacts"
+CONFIG_KEY_ARTIFACT_RELATIVE_PATH = "path"
+CONFIG_KEY_ARTIFACT_URI = "uri"
+CONFIG_KEY_PYTHON_MODEL = "python_model"
+
+
+class PythonModel(object):
+    """
+    Represents a generic Python model that evaluates inputs and produces API-compatible outputs.
+    By subclassing :class:`~PythonModel`, users can create customized MLflow models with the
+    "python_function" ("pyfunc") flavor, leveraging custom inference logic and artifact
+    dependencies.
+    """
+    __metaclass__ = ABCMeta
+
+    def load_context(self, context):
+        """
+        Loads artifacts from the specified :class:`~PythonModelContext` that can be used by
+        ``predict()`` when evaluating inputs.
+
+        :param context: A :class:`~PythonModelContext` instance containing artifacts that the model
+                        can use to perform inference.
+        """
+        # pylint: disable=attribute-defined-outside-init
+        self.context = context
```

Comment:
Hmm. I'm not sure this makes sense -- it is quite reasonable to forget to call the parent load_context method, in which case a contract such as "self.context is available" would be violated. Maybe simpler would be to pass context into `predict`, such that `load_context` becomes merely a hook for optimization.